### ISSUE TITLE
No Limit on Cache Size

### ISSUE THREAD

**Issue by tehryanx** [hotwired/turbo#922]
Using `data-turbo-preload`, it's possible to overload the cache and make the browser unstable. This turns a simple link injection bug into a full browser DOS. Below are screenshots where I've loaded a 30mb and a 50mb file into cache using nothing but a link tag. Note how the memory footprint balloons.

![image](https://user-images.githubusercontent.com/8878295/236468896-a0415816-8afb-4eae-bad0-4118ba889155.png)

![image](https://user-images.githubusercontent.com/8878295/236469322-2f5c5571-603b-49e1-ac3f-5174e20e37b8.png)

**Comment by seanpdoyle** [comment]
@tehryanx thank you for opening this issue!

I'm surprised that this is possible. The `SnapshotClass` has had a `size` property since the project was [ported from Turbolinks](https://github.com/hotwired/turbo/blame/0826b8152c0e97f19d459c1a1c364fa89cc62829/src/core/drive/snapshot_cache.ts#L19-L22). Calls to `put` and `get` on that call invoke `touch()`, which them invokes `trim()`.

Recently, the original implementation has been split into a new `MemoryStore` cache, which [still accepts a size](https://github.com/hotwired/turbo/blob/c207f5b25758e4a084e8ae42e49712b91cf37114/src/core/drive/cache_stores/memory_store.js#L7-L9), and has a default `size` of [10 entries](https://github.com/hotwired/turbo/blob/c207f5b25758e4a084e8ae42e49712b91cf37114/src/core/drive/snapshot_cache.js#L10).

@afcapel @kevinmcconnell since you two have recently implemented the new caching mechanisms, this is definitely worth investigating further.

**Comment by afcapel** [comment]
@seanpdoyle I don't think this issue is related to the disk cache additions, since it predates [the PR](https://github.com/hotwired/turbo/pull/949) being merged.

The normal Turbo cache is an in memory cache, so preloading a very big page will load it into memory. That's the expected behaviour. You can fix this removing the preloading or [opting out of caching](https://turbo.hotwired.dev/handbook/building#opting-out-of-caching).

Agreed that now that we have a disk cache it'd be interesting to explore if we can limit the cache byte size, instead of limiting the number entries, but I suspect it will be a bit difficult to nail down in a general way.
